File: packages/fragment-cli/templates/electron/src/agents/codex.ts
import { spawn } from 'child_process';
import type { AgentConnection, AgentResponse } from './types.js';
import type { ToolCall } from '{{PACKAGE_SCOPE}}/core/shared/types.js';

export class CodexAgent implements AgentConnection {
  model = 'codex' as const;
  private connected = false;

  async connect(): Promise<void> { this.connected = true; }
  async disconnect(): Promise<void> { this.connected = false; }
  isConnected(): boolean { return this.connected; }

  async sendMessage(
    content: string,
    onChunk: (text: string) => void,
    onToolCall: (toolCall: ToolCall) => void,
  ): Promise<AgentResponse> {
    return new Promise((resolve) => {
      const child = spawn('codex', ['--quiet', '-m', content], {
        stdio: ['pipe', 'pipe', 'pipe'], shell: true,
      });

      let fullContent = '';
      const toolCalls: ToolCall[] = [];

      child.stdout?.on('data', (data: Buffer) => {
        const text = data.toString();
        fullContent += text;
        onChunk(text);
      });

      child.on('close', (code) => {
        if (code !== 0 && !fullContent) {
          fullContent = `Codex exited with code ${code}. Ensure codex CLI is installed.`;
          onChunk(fullContent);
        }
        resolve({ content: fullContent, toolCalls });
      });

      child.on('error', (err) => {
        const msg = `Failed to start codex: ${err.message}. Ensure codex CLI is installed.`;
        onChunk(msg);
        resolve({ content: msg, toolCalls: [] });
      });
    });
  }
}

File: packages/fragment-cli/templates/tui/agent/codex.go
package agent

import (
	"bytes"
	"os/exec"
)

// CodexAgent connects to Codex via CLI subprocess.
type CodexAgent struct {
	connected bool
}

func NewCodexAgent() *CodexAgent {
	return &CodexAgent{}
}

func (a *CodexAgent) Model() string { return "codex" }

func (a *CodexAgent) Connect() error {
	a.connected = true
	return nil
}

func (a *CodexAgent) Disconnect() {
	a.connected = false
}

func (a *CodexAgent) IsConnected() bool {
	return a.connected
}

func (a *CodexAgent) SendMessage(content string, onChunk func(string)) (string, error) {
	cmd := exec.Command("codex", "--quiet", "-m", content)
	var out bytes.Buffer
	cmd.Stdout = &out

	if err := cmd.Run(); err != nil {
		msg := "Codex CLI not available. Ensure it is installed."
		onChunk(msg)
		return msg, nil
	}

	result := out.String()
	onChunk(result)
	return result, nil
}
